frontend/src/metabase/core/components/Select/Select.tsx

```diff
@@ -16,70 +24,103 @@ import { SelectAccordionList } from "./Select.styled";
+  defaultValue?: TValue;
+  onChange: (event: SelectChangeEvent<TValue>) => void;
+  multiple?: boolean;
+  placeholder?: string;
+  disabled?: boolean;
+  hiddenIcons?: boolean;
+
+  // PopoverWithTrigger props
+  isInitiallyOpen?: boolean;
+  triggerElement?: ReactNode;
+  onClose?: () => void;
+
+  // SelectButton props
+  buttonProps?: Partial<SelectButtonProps>;
+  buttonText?: string; // will override selected options text
+
+  // AccordionList props
+  searchProp?: string;
+  searchCaseInsensitive?: boolean;
+  searchPlaceholder?: string;
+  searchFuzzy?: boolean;
+  hideEmptySectionsInSearch?: boolean;
+  width?: number;
+
+  optionNameFn?: (option: TOption) => string | undefined;
+  optionValueFn?: (option: TOption) => TValue;
+  optionDescriptionFn?: (option: TOption) => string | undefined;
+  optionSectionFn?: (option: TOption) => string;
+  optionDisabledFn?: (option: TOption) => boolean;
+  optionIconFn?: (option: TOption) => string | undefined;
+  optionClassNameFn?: (option: TOption) => string | undefined;
+  optionStylesFn?: (option: TOption) => CSSProperties | undefined;
+
+  footer?: ReactNode;
+}
+
+export interface SelectOption<TValue = Key> {
+  value: TValue;
+  name?: string;
+  description?: string;
+  icon?: string;
+  iconSize?: number;
+  iconColor?: string;
+  disabled?: boolean;
+  children?: ReactNode;
+}
+
+export interface SelectSection<TOption = SelectOption> {
+  name?: string;
+  icon?: string;
+  items: TOption[];
+}
+
+export interface SelectChangeEvent<TValue> {
+  target: SelectChangeTarget<TValue>;
+}
+
+export interface SelectChangeTarget<TValue> {
+  name?: string;
+  value: TValue;
+}
+
+class Select<TValue, TOption = SelectOption<TValue>> extends Component<
+  SelectProps<TValue, TOption>
+> {
+  _popover?: any;
+  selectButtonRef: RefObject<any>;
+  _getValues: () => TValue[];
+  _getValuesSet: () => Set<TValue>;
 
   static defaultProps = {
-    optionNameFn: option => option.children || option.name,
-    optionValueFn: option => option.value,
-    optionDescriptionFn: option => option.description,
-    optionDisabledFn: option => option.disabled,
-    optionIconFn: option => option.icon,
+    optionNameFn: (option: SelectOption) => option.children || option.name,
+    optionValueFn: (option: SelectOption) => option.value,
+    optionDescriptionFn: (option: SelectOption) => option.description,
+    optionDisabledFn: (option: SelectOption) => option.disabled,
+    optionIconFn: (option: SelectOption) => option.icon,
   };
 
-  constructor(props) {
+  constructor(props: SelectProps<TValue, TOption>) {
     super(props);
 
     // reselect selectors
-    const _getValue = props =>
+    const _getValue = (props: SelectProps<TValue, TOption>) =>
       // If a defaultValue is passed, replace a null value with it.
       // Otherwise, allow null values since we sometimes want them.
       Object.prototype.hasOwnProperty.call(props, "defaultValue") &&
-      props.value == null
+      props.value == null &&
```

Comment:
Ok I promised that I haven't changed anything in Select but in fact that's the only line actually changed for types to work correctly. Please double check this but I shouldn't change the behavior. Previously, when `value == null` we used the `defaultValue`, but now we do that only when `defaultValue` itself is `!= null`. So if both `value` and `defaultValue` are `== null`, there should be no difference between `value` and `defaultValue`, unless we treat `null` and `undefined` differently in that case somewhere.